ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java

```diff
@@ -61,20 +63,45 @@
 
   private final Rewriter<C> rewriter;
 
+  public static final class Context<C> {
+    private final C context;
+    private final Rewriter<C> rewriter;
+
+    private Context(final C context, final Rewriter<C> rewriter) {
+      this.context = context;
+      this.rewriter = rewriter;
+    }
+
+    public C getContext() {
+      return context;
+    }
+
+    public AstNode process(final AstNode node) {
+      return rewriter.process(node, context);
+    }
+
+    public Expression process(final Expression expression) {
+      return rewriter.processExpression(expression, context);
+    }
+  }
+
   /**
    * Creates a new StatementRewriter that rewrites all expressions in a statement by
    * using the provided expression-rewriter.
    * @param expressionRewriter The expression rewriter used to rewrite an expression.
    */
-  StatementRewriter(final BiFunction<Expression, C, Expression> expressionRewriter) {
-    this.rewriter = new Rewriter<>(expressionRewriter);
+  StatementRewriter(
+      final BiFunction<Expression, C, Expression> expressionRewriter,
+      final BiFunction<AstNode, Context<C>, Optional<AstNode>> plugin) {
```

Comment:
update java doc with `plugin`. It's a new term / pattern. It would be good to explain what the plugin is and how its used.  It seems that it gets first dibs are rewriting things...